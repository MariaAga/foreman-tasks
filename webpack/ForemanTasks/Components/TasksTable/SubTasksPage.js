import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const SubTasksPage = props => {
  const getBreadcrumbs = actionName => ({
    breadcrumbItems: [
      { caption: __('Tasks'), url: `/foreman_tasks/tasks` },
      {
        caption: actionName,
        url: `/foreman_tasks/tasks/${props.match.params.id}`,
      },
      { caption: __('Sub tasks') },
    ],
  });
  return <TasksTablePage getBreadcrumbs={getBreadcrumbs} {...props} />;
};

SubTasksPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.object,
  }),
};

SubTasksPage.defaultProps = {
  match: {
    params: {},
  },
};
