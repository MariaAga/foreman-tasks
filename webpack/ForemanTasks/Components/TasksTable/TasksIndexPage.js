import React from 'react';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const TasksIndexPage = props => {
  const getBreadcrumbs = () => ({
    breadcrumbItems: [{ caption: __('Tasks'), url: `/foreman_tasks/tasks` }],
  });
  return <TasksTablePage getBreadcrumbs={getBreadcrumbs} {...props} />;
};
