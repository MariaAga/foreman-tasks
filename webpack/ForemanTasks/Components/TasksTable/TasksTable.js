import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Table } from 'foremanReact/components/common/table';
import { STATUS } from 'foremanReact/constants';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { translate as __ } from 'foremanReact/common/I18n';
import Pagination from 'foremanReact/components/Pagination/PaginationWrapper';
import createTasksTableSchema from './TasksTableSchema';

const TasksTable = ({
  getTableItems,
  error,
  status,
  results,
  query,
  itemCount,
  pagination,
  setSort,
  changeTablePage,
  sort,
}) => {
  useEffect(() => {
    getTableItems();
  }, [
    query.state,
    query.result,
    query.mode,
    query.time,
    pagination.page,
    pagination.perPage,
    sort.by,
    sort.order,
  ]);

  if (status === STATUS.ERROR) {
    return (
      <MessageBox
        key="tasks-table-error"
        icontype="error-circle-o"
        msg={__(`Could not receive data: ${error && error.message}`)}
      />
    );
  }

  if (status === STATUS.PENDING && results.length === 0) {
    return <div />;
  }

  if (results.length === 0) {
    return <span>{__('No Tasks')}</span>;
  }

  return (
    <div className="tasks-table">
      <Table
        key="tasks-table"
        columns={createTasksTableSchema(setSort, sort.by, sort.order)}
        rows={results}
      />
      <Pagination
        className="tasks-pagination"
        viewType="table"
        itemCount={itemCount}
        pagination={pagination}
        onChange={changeTablePage}
        dropdownButtonId="tasks-table-dropdown"
      />
    </div>
  );
};

TasksTable.propTypes = {
  results: PropTypes.array.isRequired,
  getTableItems: PropTypes.func.isRequired,
  setSort: PropTypes.func.isRequired,
  changeTablePage: PropTypes.func.isRequired,
  status: PropTypes.oneOf(Object.keys(STATUS)),
  query: PropTypes.shape({
    state: PropTypes.string,
    result: PropTypes.string,
    mode: PropTypes.string,
    time: PropTypes.string,
  }),
  error: PropTypes.instanceOf(Error),
  itemCount: PropTypes.number.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }),
  sort: PropTypes.shape({
    by: PropTypes.string,
    order: PropTypes.string,
  }).isRequired,
};

TasksTable.defaultProps = {
  status: STATUS.PENDING,
  query: {},
  error: null,
  pagination: {
    page: 1,
    perPage: 20,
  },
};

export default TasksTable;
