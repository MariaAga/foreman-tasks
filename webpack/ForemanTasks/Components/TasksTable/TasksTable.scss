.tasks-pagination {
  margin-top: -6px;
}

#tasks-table {
  margin-bottom: 70px;
}
