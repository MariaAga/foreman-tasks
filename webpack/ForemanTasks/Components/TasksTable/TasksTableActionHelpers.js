import { translate as __, sprintf } from 'foremanReact/common/I18n';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { TASKS_DASHBOARD_AVAILABLE_QUERY_MODES } from '../TasksDashboard/TasksDashboardConstants';
import { timeToHoursNumber } from '../TasksDashboard/TasksDashboardHelper';

export const convertDashboardQuery = query => {
  const {
    time_mode: timeMode,
    time_horizon: timeHorizon,
    state,
    result,
    search,
  } = query;

  const hours = timeToHoursNumber(timeHorizon);

  const timestamp = new Date(new Date() - hours * 60 * 60 * 1000);
  let dashboardTime = '';
  const stateQuery = state ? `state=${state}` : '';
  const resultQuery = result ? `result=${result}` : '';
  if (timeMode === TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.LAST) {
    dashboardTime = `(state_updated_at<${timestamp.toISOString()} or state_updated_at = NULL)`;
  } else if (timeMode === TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.OLDER) {
    dashboardTime = `(state_updated_at>${timestamp.toISOString()})`;
  }
  const newQuery = [stateQuery, resultQuery, search, dashboardTime]
    .filter(Boolean)
    .join(' and ');
  return newQuery;
};

export const resumeToastInfo = {
  resumed: { type: 'success', text: __('was resumed') },
  failed: { type: 'error', text: __('could not be resumed') },
  skipped: { type: 'warning', text: __('task has to be resumable') },
};

export const cancelToastInfo = {
  cancelled: { type: 'success', text: __('was cancelled') },
  skipped: { type: 'warning', text: __('task has to be cancellable') },
};

export const toastDispatch = ({ type, name, toastInfo, dispatch }) => {
  dispatch(
    addToast({
      type: toastInfo[type].type,
      message: sprintf('%(name)s Task execution %(type)s', {
        name,
        type: toastInfo[type].text,
      }),
    })
  );
};
