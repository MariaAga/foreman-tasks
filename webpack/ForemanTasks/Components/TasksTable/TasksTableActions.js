import { getURIQuery } from 'foremanReact/common/helpers';
import { getTableItemsAction } from 'foremanReact/components/common/table';
import API from 'foremanReact/API';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { translate as __ } from 'foremanReact/common/I18n';
import {
  TASKS_TABLE_ID,
  SELECT_ROWS,
  UNSELECT_ALL_ROWS,
  UNSELECT_ROWS,
  TASKS_TABLE_SELECTED_MODAL,
  CLOSED,
  RESUME,
  CANCEL,
} from './TasksTableConstants';
import { getApiPathname } from './TasksTableHelpers';
import { fetchTasksSummary } from '../TasksDashboard/TasksDashboardActions';

export const getTableItems = url =>
  getTableItemsAction(TASKS_TABLE_ID, getURIQuery(url), getApiPathname(url));

export const cancelTask = (id, name, url, parentTaskID) => async dispatch => {
  await dispatch(cancelTaskRequest(id, name));
  dispatch(getTableItems(url));
  dispatch(fetchTasksSummary(getURIQuery(url).time, parentTaskID));
};

export const cancelTaskRequest = (id, name) => async dispatch => {
  dispatch(
    addToast({
      type: 'info',
      message: `${__('Trying to cancel')} "${name}" ${__('task')}`,
    })
  );
  try {
    await API.post(`/foreman_tasks/tasks/${id}/cancel`);
    dispatch(
      addToast({
        type: 'success',
        message: `"${name}" ${__('Task cancelled')}`,
      })
    );
  } catch ({ response }) {
    dispatch(
      addToast({
        type: 'error',
        message: `"${name}" ${__('Task cannot be cancelled at the moment.')}`,
      })
    );
  }
};

export const resumeTask = (id, name, url, parentTaskID) => async dispatch => {
  await dispatch(resumeTaskRequest(id, name));
  dispatch(getTableItems(url));
  dispatch(fetchTasksSummary(getURIQuery(url).time), parentTaskID);
};

export const resumeTaskRequest = (id, name) => async dispatch => {
  try {
    await API.post(`/foreman_tasks/tasks/${id}/resume`);
    dispatch(
      addToast({
        type: 'success',
        message: __(`"${name}" ${__('Task execution was resumed')}`),
      })
    );
  } catch ({ response }) {
    dispatch(
      addToast({
        type: 'error',
        message: __(`Task "${name}" has to be resumable.`),
      })
    );
  }
};

export const selectAllRows = results => ({
  type: SELECT_ROWS,
  payload: results.map(row => row.id),
});

export const unselectAllRows = () => ({
  type: UNSELECT_ALL_ROWS,
});

export const selectRow = id => ({
  type: SELECT_ROWS,
  payload: [id],
});

export const unselectRow = id => ({
  type: UNSELECT_ROWS,
  payload: id,
});

export const actionSelected = (actionType, selected, url) => async dispatch => {
  let notAllActionable = false;
  let someActionable = false;
  const promises = selected.map(task => {
    if (actionType === RESUME && task.isResumeble) {
      someActionable = true;
      return dispatch(resumeTaskRequest(task.id, task.name, url));
    } else if (actionType === CANCEL && task.isCancelleble) {
      someActionable = true;
      return dispatch(cancelTaskRequest(task.id, task.name, url));
    }
    notAllActionable = true;
    return null;
  });
  if (notAllActionable)
    dispatch(
      addToast({
        type: 'warning',
        message: __(
          `Not all the selected tasks can be ${
            actionType === RESUME ? 'resumed' : 'canceled'
          }`
        ),
      })
    );
  if (someActionable) {
    await Promise.all(promises);
    dispatch(getTableItems(url));
    dispatch(fetchTasksSummary(getURIQuery(url).time));
  }
};

export const showCancelSelcetedModal = () => ({
  type: TASKS_TABLE_SELECTED_MODAL,
  payload: CANCEL,
});

export const showResumeSelcetedModal = () => ({
  type: TASKS_TABLE_SELECTED_MODAL,
  payload: RESUME,
});

export const hideSelcetedModal = () => ({
  type: TASKS_TABLE_SELECTED_MODAL,
  payload: CLOSED,
});
