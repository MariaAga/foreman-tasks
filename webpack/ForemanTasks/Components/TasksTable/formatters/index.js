export { default as selectionHeaderCellFormatter } from './selectionHeaderCellFormatter';
export { default as selectionCellFormatter } from './selectionCellFormatter';

export { actionNameCellFormatter } from './actionNameCellFormatter';
export { actionCellFormatter } from './actionCellFormatter';
export { dateCellFormmatter } from './dateCellFormmatter';
export { durationCellFormmatter } from './durationCellFormmatter';
