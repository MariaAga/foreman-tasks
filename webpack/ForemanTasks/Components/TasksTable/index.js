import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TasksTable from './TasksTable';
import reducer from './TasksTableReducer';
import * as actions from './TasksTableActions';
import { selectQuery } from '../TasksDashboard/TasksDashboardSelectors';
import {
  selectStatus,
  selectError,
  selectResults,
  selectPagitation,
  selectItemCount,
  selectSort,
} from './TasksTableSelectors';

const mapStateToProps = state => ({
  status: selectStatus(state),
  error: selectError(state),
  sort: selectSort(state),
  results: selectResults(state),
  query: selectQuery(state),
  pagination: selectPagitation(state),
  itemCount: selectItemCount(state),
});
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { tasksTable: reducer };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TasksTable);
