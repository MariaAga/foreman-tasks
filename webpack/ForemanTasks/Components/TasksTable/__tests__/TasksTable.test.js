import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import { STATUS } from 'foremanReact/constants';
import TasksTable from '../TasksTable';

const actions = {
  getTableItems: jest.fn(),
  setSort: jest.fn(),
  changeTablePage: jest.fn(),
};

const fixtures = {
  'render with minimal Props': {
    ...actions,
    status: STATUS.RESOLVED,
    results: ['a'],
    itemCount: 0,
    pagination: {
      page: 1,
      perPage: 10,
    },
    sort: {
      by: 'q',
      order: 'w',
    },
  },
  'render with no results': {
    ...actions,
    results: [],
    status: STATUS.RESOLVED,
    itemCount: 0,
    pagination: {
      page: 1,
      perPage: 10,
    },
    sort: {
      by: 'q',
      order: 'w',
    },
  },
  'render with error Props': {
    ...actions,
    results: ['a'],
    status: STATUS.ERROR,
    itemCount: 0,
    pagination: {
      page: 1,
      perPage: 10,
    },
    sort: {
      by: 'q',
      order: 'w',
    },
  },
  'render with loading Props': {
    ...actions,
    results: ['a'],
    status: STATUS.PENDING,
    itemCount: 0,
    pagination: {
      page: 1,
      perPage: 10,
    },
    sort: {
      by: 'q',
      order: 'w',
    },
  },
};

jest.mock('../TasksTableSchema');
describe('TasksTable', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TasksTable, fixtures));
});
