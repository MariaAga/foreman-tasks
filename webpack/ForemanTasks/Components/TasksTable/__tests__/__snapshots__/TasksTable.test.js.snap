// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksTable rendering render with error Props 1`] = `
<MessageBox
  icontype="error-circle-o"
  key="tasks-table-error"
  msg="Could not receive data: null"
/>
`;

exports[`TasksTable rendering render with loading Props 1`] = `
<div
  className="tasks-table"
>
  <Table
    key="tasks-table"
    rows={
      Array [
        "a",
      ]
    }
  />
  <PaginationWrapper
    className="tasks-pagination"
    dropdownButtonId="tasks-table-dropdown"
    itemCount={2}
    onChange={[MockFunction]}
    pagination={
      Object {
        "page": 1,
        "perPage": 10,
      }
    }
    viewType="table"
  />
</div>
`;

exports[`TasksTable rendering render with minimal Props 1`] = `
<div
  className="tasks-table"
>
  <Table
    key="tasks-table"
    rows={
      Array [
        "a",
        "b",
      ]
    }
  />
  <PaginationWrapper
    className="tasks-pagination"
    dropdownButtonId="tasks-table-dropdown"
    itemCount={2}
    onChange={[MockFunction]}
    pagination={
      Object {
        "page": 1,
        "perPage": 10,
      }
    }
    viewType="table"
  />
</div>
`;

exports[`TasksTable rendering render with no results 1`] = `
<span>
  No Tasks
</span>
`;
